ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java

```diff
@@ -140,81 +131,38 @@ private void validateIndex(int index) {
   static class BlobReadInfo {
     private final CloudBlobMetadata blobMetadata;
     private final BlobId blobId;
-    private ByteBuffer prefetchedBuffer;
-    private int position = -1;
-    private int size = -1;
-    private boolean isPrefetched;
+    private ByteBuf blobContent = null;
 
     public BlobReadInfo(CloudBlobMetadata blobMetadata, BlobId blobId) {
       this.blobMetadata = blobMetadata;
       this.blobId = blobId;
-      this.isPrefetched = false;
     }
 
     /**
-     * Prefetch the {@code blob} from {@code CloudDestination} and put it in {@code prefetchedBuffer}
+     * Download the {@code blob} from {@code CloudDestination} and put it in {@code blobContent}
      * @param blobStore {@code CloudBlobStore} implementation representing the cloud from which download will happen.
      * @throws StoreException if blob cloud not be downloaded
      */
-    public void prefetchBlob(CloudBlobStore blobStore) throws StoreException {
+    public void downloadBlob(CloudBlobStore blobStore) throws StoreException {
       // Casting blobsize to int, as blobs are chunked in Ambry, and chunk size is 4/8MB.
       // However, if in future, if very large size of blobs are allowed, then prefetching logic should be changed.
-      prefetchedBuffer = ByteBuffer.allocate((int) blobMetadata.getSize());
-      ByteBufferOutputStream outputStream = new ByteBufferOutputStream(prefetchedBuffer);
-      blobStore.downloadBlob(blobMetadata, blobId, outputStream);
-      prefetchedBuffer.flip();
-      isPrefetched = true;
-    }
-
-    /**
-     * Donwload the blob from the {@code CloudDestination} to the {@code OutputStream}
-     * @param blobStore {@code CloudBlobStore} implementation representing the cloud from which download will happen.
-     * @param outputStream OutputStream to download the blob to
-     * @throws StoreException if blob download fails.
-     */
-    public void downloadBlob(CloudBlobStore blobStore, OutputStream outputStream) throws StoreException {
+      blobContent = PooledByteBufAllocator.DEFAULT.ioBuffer((int) blobMetadata.getSize());
```

Comment:
is it safe to use netty ByteBuf in CloudMessageReadSet? @ankagrawal  what is the cloud network implementation? does it use http2? socket server?

Reply:
@justinlin-linkedin you are concerning if it's used in socket server, the ByteBuf can't be released?  
storeMessageSet had the same problem. I guess we should retire socket server soon.

Reply:
For vcr the network implementation currently is socket server based. It doesn't use http2.

Reply:
For now, until we have netty based http2 implementation in cloud, is there an issue with using `ByteBuffer.allocate` for cloud?

Reply:
Both socket and http network stack can take care of allocated ByteBuf. As long as we have no cloud based network implementation, it's good. (Or release ByteBuf correctly in new implementations)